pkg/helm/watcher/controller/helm_watcher_test.go

```diff
@@ -63,7 +63,9 @@ func TestReconcile(t *testing.T) {
 	fakeClient := fake.NewClientBuilder().WithScheme(scheme).WithObjects(repo)
 
 	fakeRepoManager.ListChartsReturns([]*pb.Profile{profile1, profile2}, nil)
-	fakeRepoManager.GetValuesFileReturns([]byte("value"), nil)
+	fakeRepoManager.GetValuesFileReturnsOnCall(0, []byte("value1"), nil)
+	fakeRepoManager.GetValuesFileReturnsOnCall(1, []byte("value2"), nil)
+	fakeRepoManager.GetValuesFileReturnsOnCall(2, []byte("value3"), nil)
```

Comment:
nit: are we validating that the `GetValuesFileArgsForCall(0,1,2)` are all the correct values?

Reply:
What do you mean? Other than this part?

```go
assert.Equal(t, expectedData, cacheData)
```

This is validating that the data arrived and is the thing that it returned where expected data is set to the multiple versions. Or do you mean to somehow validate the mock?

Reply:
We aren't validating that the calls to the `GetValuesFiles` are containing the right values. E.g. You could make the args provided her gibberish, and the tests would pass: https://github.com/weaveworks/weave-gitops/blob/eb8c7f779b158272251312d3f8c010cd2e508973/pkg/helm/watcher/controller/helm_watcher.go#L80

We should add an assertion that:
```
_, repo, chartRef,fileName := fakeRepoManager.GetValuesFileArgsForCall(0)
Expect(repo).To(Equal(....))
Expect(chartRef).To(Equal(....))
Expect(fileName).To(Equal(....))

```

Reply:
No, sorry, I'm not sure I'm following. :DDD 

If I mess it up it will fail:
```
        	            	   (string) (len=15) "test-profiles-2": (map[string][]uint8) (len=1) {
        	            	-   (string) (len=5) "0.0.4": ([]uint8) (len=4) {
        	            	-    00000000  61 73 64 66                                       |asdf|
        	            	+   (string) (len=5) "0.0.4": ([]uint8) (len=6) {
        	            	+    00000000  76 61 6c 75 65 33                                 |value3|
        	            	    }
```

Reply:
```go
	expectedData := cache.Data{
		Profiles: []*pb.Profile{profile1, profile2},
		Values: map[string]map[string][]byte{
			profile1.Name: {
				"0.0.1": []byte("value1"),
				"0.0.2": []byte("value2"),
			},
			profile2.Name: {
				"0.0.4": []byte("asdf"),
			},
		},
	}
	_, namespace, name, cacheData := fakeCache.PutArgsForCall(0)
	assert.Equal(t, "test-namespace", namespace)
	assert.Equal(t, "test-name", name)
	assert.Equal(t, expectedData, cacheData)
```

If the expected data here is not the thing that the above call returns than it fails comparing it to the expected data.

Reply:
If i modify the real code:
```
            valueBytes, err := r.RepoManager.GetValuesFile(context.Background(), &repository, &helm.ChartReference{
                Chart:   chart.Name,
                Version: v,
            }, chartutil.ValuesfileName)
            
      >>>>>>>
            valueBytes, err := r.RepoManager.GetValuesFile(context.Background(), nil, &helm.ChartReference{}, "foo")
```

and run the tests, it passes:
```
ginkgo -r pkg/helm/watcher/controller
=== RUN   TestArtifactUpdatePredicate_Update
=== RUN   TestArtifactUpdatePredicate_Update/returns_false_no_new_or_old_object_detected
=== RUN   TestArtifactUpdatePredicate_Update/returns_false_if_old_source_is_not_a_sourcev1.Source_object
=== RUN   TestArtifactUpdatePredicate_Update/returns_false_if_new_source_is_not_a_sourcev1.Source_object
=== RUN   TestArtifactUpdatePredicate_Update/returns_true_if_old_source_does_not_have_an_artifact_but_new_does
=== RUN   TestArtifactUpdatePredicate_Update/returns_true_if_old_source_and_new_source_artifact_revision_doesn't_match
=== RUN   TestArtifactUpdatePredicate_Update/returns_false_if_old_and_new_artifact_revision_are_the_same
--- PASS: TestArtifactUpdatePredicate_Update (0.00s)
    --- PASS: TestArtifactUpdatePredicate_Update/returns_false_no_new_or_old_object_detected (0.00s)
    --- PASS: TestArtifactUpdatePredicate_Update/returns_false_if_old_source_is_not_a_sourcev1.Source_object (0.00s)
    --- PASS: TestArtifactUpdatePredicate_Update/returns_false_if_new_source_is_not_a_sourcev1.Source_object (0.00s)
    --- PASS: TestArtifactUpdatePredicate_Update/returns_true_if_old_source_does_not_have_an_artifact_but_new_does (0.00s)
    --- PASS: TestArtifactUpdatePredicate_Update/returns_true_if_old_source_and_new_source_artifact_revision_doesn't_match (0.00s)
    --- PASS: TestArtifactUpdatePredicate_Update/returns_false_if_old_and_new_artifact_revision_are_the_same (0.00s)
=== RUN   TestGenerationUpdateReconcilerPredicate_Update
=== RUN   TestGenerationUpdateReconcilerPredicate_Update/returns_false_no_new_or_old_object_detected
=== RUN   TestGenerationUpdateReconcilerPredicate_Update/returns_false_no_old
=== RUN   TestGenerationUpdateReconcilerPredicate_Update/returns_false_no_new
=== RUN   TestGenerationUpdateReconcilerPredicate_Update/returns_false_if_old_source's_generation_is_lower_than_the_new_source's_but_no_deletion_ts
=== RUN   TestGenerationUpdateReconcilerPredicate_Update/returns_true_if_old_source's_generation_is_lower_than_the_new_source's_and_deletion_ts_is_set
=== RUN   TestGenerationUpdateReconcilerPredicate_Update/returns_false_if_old_source's_generation_is_higher_than_the_new_source's
=== RUN   TestGenerationUpdateReconcilerPredicate_Update/returns_false_if_old_source's_generation_equals_with_the_new_source's
--- PASS: TestGenerationUpdateReconcilerPredicate_Update (0.00s)
    --- PASS: TestGenerationUpdateReconcilerPredicate_Update/returns_false_no_new_or_old_object_detected (0.00s)
    --- PASS: TestGenerationUpdateReconcilerPredicate_Update/returns_false_no_old (0.00s)
    --- PASS: TestGenerationUpdateReconcilerPredicate_Update/returns_false_no_new (0.00s)
    --- PASS: TestGenerationUpdateReconcilerPredicate_Update/returns_false_if_old_source's_generation_is_lower_than_the_new_source's_but_no_deletion_ts (0.00s)
    --- PASS: TestGenerationUpdateReconcilerPredicate_Update/returns_true_if_old_source's_generation_is_lower_than_the_new_source's_and_deletion_ts_is_set (0.00s)
    --- PASS: TestGenerationUpdateReconcilerPredicate_Update/returns_false_if_old_source's_generation_is_higher_than_the_new_source's (0.00s)
    --- PASS: TestGenerationUpdateReconcilerPredicate_Update/returns_false_if_old_source's_generation_equals_with_the_new_source's (0.00s)
=== RUN   TestReconcile
--- PASS: TestReconcile (0.00s)
=== RUN   TestReconcileGetChartFails
--- PASS: TestReconcileGetChartFails (0.00s)
=== RUN   TestReconcileGetValuesFileFailsItWillContinue
--- PASS: TestReconcileGetValuesFileFailsItWillContinue (0.00s)
=== RUN   TestReconcileIgnoreReposWithoutArtifact
--- PASS: TestReconcileIgnoreReposWithoutArtifact (0.00s)
=== RUN   TestReconcileUpdateReturnsError
--- PASS: TestReconcileUpdateReturnsError (0.00s)
=== RUN   TestReconcileKubernetesGetFails
--- PASS: TestReconcileKubernetesGetFails (0.00s)
PASS

Ginkgo ran 1 suite in 3.965862719s
Test Suite Passed
```

Reply:
Oh you mean THAT! Ahh, now I see. :D I don't care about that part. :D GetValuesFromFile is tested in chart_test.

Reply:
right, but your not testing in `helm_watcher.go` that it invokes `GetValuesFromFile` correctly? I could change this line, and merge the PR and nothing would fail until the acceptance tests

Reply:
nothing fails inside `pkg/helm` with that modification:
```
git diff
            valueBytes, err := r.RepoManager.GetValuesFile(context.Background(), &repository, &helm.ChartReference{
                Chart:   chart.Name,
                Version: v,
            }, chartutil.ValuesfileName)
            ----
            valueBytes, err := r.RepoManager.GetValuesFile(context.Background(), nil, &helm.ChartReference{}, "foo")

```

```
ginkgo -r pkg/helm/
[1642679993] Charts Suite - 14/14 specs •••••••••••••• SUCCESS! 21.284035ms PASS
=== RUN   TestCacheListProfiles
--- PASS: TestCacheListProfiles (0.00s)
=== RUN   TestCacheListProfilesNotFound
--- PASS: TestCacheListProfilesNotFound (0.00s)
=== RUN   TestCacheListProfilesInvalidDataInFile
--- PASS: TestCacheListProfilesInvalidDataInFile (0.00s)
=== RUN   TestCacheGetProfileValues
--- PASS: TestCacheGetProfileValues (0.00s)
=== RUN   TestGetProfileValuesNonexistent
--- PASS: TestGetProfileValuesNonexistent (0.00s)
=== RUN   TestDeleteExistingData
--- PASS: TestDeleteExistingData (0.00s)
=== RUN   TestListProfilesFailedLock
--- PASS: TestListProfilesFailedLock (0.00s)
=== RUN   TestGetProfileValuesFailedLock
--- PASS: TestGetProfileValuesFailedLock (0.00s)
=== RUN   TestUpdateFailedLock
--- PASS: TestUpdateFailedLock (0.00s)
=== RUN   TestDeleteFailedLock
--- PASS: TestDeleteFailedLock (0.00s)
PASS
=== RUN   TestArtifactUpdatePredicate_Update
=== RUN   TestArtifactUpdatePredicate_Update/returns_false_no_new_or_old_object_detected
=== RUN   TestArtifactUpdatePredicate_Update/returns_false_if_old_source_is_not_a_sourcev1.Source_object
=== RUN   TestArtifactUpdatePredicate_Update/returns_false_if_new_source_is_not_a_sourcev1.Source_object
=== RUN   TestArtifactUpdatePredicate_Update/returns_true_if_old_source_does_not_have_an_artifact_but_new_does
=== RUN   TestArtifactUpdatePredicate_Update/returns_true_if_old_source_and_new_source_artifact_revision_doesn't_match
=== RUN   TestArtifactUpdatePredicate_Update/returns_false_if_old_and_new_artifact_revision_are_the_same
--- PASS: TestArtifactUpdatePredicate_Update (0.00s)
    --- PASS: TestArtifactUpdatePredicate_Update/returns_false_no_new_or_old_object_detected (0.00s)
    --- PASS: TestArtifactUpdatePredicate_Update/returns_false_if_old_source_is_not_a_sourcev1.Source_object (0.00s)
    --- PASS: TestArtifactUpdatePredicate_Update/returns_false_if_new_source_is_not_a_sourcev1.Source_object (0.00s)
    --- PASS: TestArtifactUpdatePredicate_Update/returns_true_if_old_source_does_not_have_an_artifact_but_new_does (0.00s)
    --- PASS: TestArtifactUpdatePredicate_Update/returns_true_if_old_source_and_new_source_artifact_revision_doesn't_match (0.00s)
    --- PASS: TestArtifactUpdatePredicate_Update/returns_false_if_old_and_new_artifact_revision_are_the_same (0.00s)
=== RUN   TestGenerationUpdateReconcilerPredicate_Update
=== RUN   TestGenerationUpdateReconcilerPredicate_Update/returns_false_no_new_or_old_object_detected
=== RUN   TestGenerationUpdateReconcilerPredicate_Update/returns_false_no_old
=== RUN   TestGenerationUpdateReconcilerPredicate_Update/returns_false_no_new
=== RUN   TestGenerationUpdateReconcilerPredicate_Update/returns_false_if_old_source's_generation_is_lower_than_the_new_source's_but_no_deletion_ts
=== RUN   TestGenerationUpdateReconcilerPredicate_Update/returns_true_if_old_source's_generation_is_lower_than_the_new_source's_and_deletion_ts_is_set
=== RUN   TestGenerationUpdateReconcilerPredicate_Update/returns_false_if_old_source's_generation_is_higher_than_the_new_source's
=== RUN   TestGenerationUpdateReconcilerPredicate_Update/returns_false_if_old_source's_generation_equals_with_the_new_source's
--- PASS: TestGenerationUpdateReconcilerPredicate_Update (0.00s)
    --- PASS: TestGenerationUpdateReconcilerPredicate_Update/returns_false_no_new_or_old_object_detected (0.00s)
    --- PASS: TestGenerationUpdateReconcilerPredicate_Update/returns_false_no_old (0.00s)
    --- PASS: TestGenerationUpdateReconcilerPredicate_Update/returns_false_no_new (0.00s)
    --- PASS: TestGenerationUpdateReconcilerPredicate_Update/returns_false_if_old_source's_generation_is_lower_than_the_new_source's_but_no_deletion_ts (0.00s)
    --- PASS: TestGenerationUpdateReconcilerPredicate_Update/returns_true_if_old_source's_generation_is_lower_than_the_new_source's_and_deletion_ts_is_set (0.00s)
    --- PASS: TestGenerationUpdateReconcilerPredicate_Update/returns_false_if_old_source's_generation_is_higher_than_the_new_source's (0.00s)
    --- PASS: TestGenerationUpdateReconcilerPredicate_Update/returns_false_if_old_source's_generation_equals_with_the_new_source's (0.00s)
=== RUN   TestReconcile
--- PASS: TestReconcile (0.00s)
=== RUN   TestReconcileGetChartFails
--- PASS: TestReconcileGetChartFails (0.00s)
=== RUN   TestReconcileGetValuesFileFailsItWillContinue
--- PASS: TestReconcileGetValuesFileFailsItWillContinue (0.00s)
=== RUN   TestReconcileIgnoreReposWithoutArtifact
--- PASS: TestReconcileIgnoreReposWithoutArtifact (0.00s)
=== RUN   TestReconcileUpdateReturnsError
--- PASS: TestReconcileUpdateReturnsError (0.00s)
=== RUN   TestReconcileKubernetesGetFails
--- PASS: TestReconcileKubernetesGetFails (0.00s)
PASS

Ginkgo ran 3 suites in 5.447425853s
Test Suite Passed
```

Reply:
finally after I understood what you wanted. :D I added the verification :)